### PR TITLE
Test Cleanup Precision issues

### PR DESCRIPTION
**Pull request by DN6**
# What does this PR do?
Moves more tests to a cosine distance based metric to assess the generated outputs of pipelines based on expected outputs. 

Approach has been discussed previsouly here: #4759 

This should reduce the number of failures seen in the Slow Push Tests from 56 to ~25-30. The remaining set of failing tests seem to be related to the following issues and pipelines and will be addressed in a separate PR. 

1. Out Of Memory Errors
2. Issues with the AudioLDM2 Pipeline
3. Issues with  the MusicLDM Pipeline
4. LoRA loading issues 
5. Precision Issues (these appear to be legitimate and not due to GPU Nondeterminism. I need to look into these further to confirm what is happening)   

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: tests/pipelines/test_pipelines_common.py [hunk 1]
```diff
@@ -329,7 +329,7 @@ def test_save_load_local(self, expected_max_difference=5e-4):
         inputs = self.get_dummy_inputs(torch_device)
         output_loaded = pipe_loaded(**inputs)[0]
 
-        max_diff = np.abs(to_np(output) - to_np(output_loaded)).max()
+        max_diff = numpy_cosine_similarity_distance(to_np(output).flatten(), to_np(output_loaded).flatten())
```
Comment (patrickvonplaten) [inline tests/pipelines/test_pipelines_common.py:332]: Same here, let's try to use this instead: https://github.com/huggingface/diffusers/pull/4812/files#r1307507791

File: tests/pipelines/test_pipelines_common.py [hunk 2]
```diff
@@ -616,7 +615,7 @@ def test_save_load_optional_components(self, expected_max_difference=1e-4):
...
```
Comment (patrickvonplaten) [inline tests/pipelines/test_pipelines_common.py:618]: https://github.com/huggingface/diffusers/pull/4812/files#r1307509747 would be better

File: tests/pipelines/test_pipelines_common.py [hunk 3]
```diff
@@ -673,7 +672,9 @@ def _test_attention_slicing_forward_pass(
         output_with_slicing = pipe(**inputs)[0]
 
         if test_max_difference:
-            max_diff = np.abs(to_np(output_with_slicing) - to_np(output_without_slicing)).max()
```
Comment (patrickvonplaten) [inline tests/pipelines/test_pipelines_common.py:676]: https://github.com/huggingface/diffusers/pull/4812/files#r1307509747 would be better

File: tests/pipelines/test_pipelines_common.py [hunk 4]
```diff
@@ -696,7 +697,9 @@ def test_cpu_offload_forward_pass(self, expected_max_diff=1e-4):
         inputs = self.get_dummy_inputs(torch_device)
         output_with_offload = pipe(**inputs)[0]
 
-        max_diff = np.abs(to_np(output_with_offload) - to_np(output_without_offload)).max()
+        max_diff = numpy_cosine_similarity_distance(
```
Comment (patrickvonplaten) [inline tests/pipelines/test_pipelines_common.py:700]: https://github.com/huggingface/diffusers/pull/4812/files#r1307509747 would be better

File: tests/pipelines/test_pipelines_common.py [hunk 5]
```diff
@@ -731,7 +734,9 @@ def _test_xformers_attention_forwardGenerator_pass(
         )
 
         if test_max_difference:
-            max_diff = np.abs(output_with_offload - output_without_offload).max()
+            max_diff = numpy_cosine_similarity_distance(
+                to_np(output_with_offload).flatten(), to_np(output_without_offload).flatten()
```
Comment (patrickvonplaten) [inline tests/pipelines/test_pipelines_common.py:738]: https://github.com/huggingface/diffusers/pull/4812/files#r1307509747 would be better

File: tests/pipelines/stable_diffusion/test_stable_diffusion.py [hunk 6]
```diff
@@ -754,7 +755,8 @@ def test_stable_diffusion_attention_slicing(self):
         # make sure that more than 3.75 GB is allocated
         mem_bytes = torch.cuda.max_memory_allocated()
         assert mem_bytes > 3.75 * 10**9
-        assert np.abs(image_sliced - image).max() < 1e-3
+        max_diff = numpy_cosine_similarity_distance(image_sliced.flatten(), image.flatten())
```
Comment (patrickvonplaten) [inline tests/pipelines/stable_diffusion/test_stable_diffusion.py:758]: Ok!

File: tests/models/test_models_unet_2d_condition.py [hunk 7]
```diff
@@ -673,7 +673,7 @@ def test_lora_save_torch_force_load_safetensors_error(self):
                 new_model.load_attn_procs(tmpdirname, use_safetensors=True)
             self.assertIn("Error no file named pytorch_lora_weights.safetensors", str(e.exception))
 
-    def test_lora_on_off(self):
+    def test_lora_on_off(self, expected_max_diff=1e-4):
```
Comment (patrickvonplaten) [inline tests/models/test_models_unet_2d_condition.py:676]: We should not need the `numpy_cosine_similarity_distance` in the model tests. Model test just test a single forward pass for which abs tensor comparison should be good enough. Let's maybe try with `1e-3` here - if this doesn't work, I'm happy to dive a bit deeper to see what's going on

**Review by patrickvonplaten — COMMENTED** [review]
Cool that more slow tests are passing with this PR. We should however be careful to not add such "easier" ways of comparison for the fast tests (those in `test_pipelines_common.py` and `tests/models/test_models_unet_2d_condition.py`). 

Fast tests **should** pass with high precision - otherwise something else might be the reason here. 

For slow tests of whole pipelines (those that test a "real" checkpoint), it's totally ok for me to use `numpy_cosine_similarity_distance`

**Comment by DN6** [comment]
Understood. Will update.

**Review by patrickvonplaten — APPROVED** [review]
Cool!

**Comment by DN6** [comment]
Hmm @patrickvonplaten I'm see a lot of failures in the fast pipeline tests using 
```python
for component in pipe.components.values():
            if hasattr(component, "set_default_attn_processor"):
                component.set_default_attn_processor()
```

```
=== 76 failed, 1314 passed, 741 skipped, 1469 warnings in 1015.78s (0:16:55) ===
```

They all seem to be coming from `lora.py`. Here's an example traceback

```
tests/pipelines/unclip/test_unclip_image_variation.py:480:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/pipelines/test_pipelines_common.py:314: in test_save_load_local
    output = pipe(**inputs)[0]
/opt/venv/lib/python3.8/site-packages/torch/utils/_contextlib.py:115: in decorate_context
    return func(*args, **kwargs)
src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:314: in __call__
    noise_pred = self.decoder(
/opt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py:1501: in _call_impl
    return forward_call(*args, **kwargs)
src/diffusers/models/unet_2d_condition.py:930: in forward
    sample, res_samples = downsample_block(
/opt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py:1501: in _call_impl
    return forward_call(*args, **kwargs)
src/diffusers/models/unet_2d_blocks.py:1725: in forward
    hidden_states = attn(
/opt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py:1501: in _call_impl
    return forward_call(*args, **kwargs)
src/diffusers/models/attention_processor.py:419: in forward
    return self.processor(
src/diffusers/models/attention_processor.py:580: in __call__
    key = attn.to_k(encoder_hidden_states, lora_scale=scale)
/opt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py:1501: in _call_impl
    return forward_call(*args, **kwargs)
src/diffusers/models/lora.py:117: in forward
    return super().forward(hidden_states)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = LoRACompatibleLinear(in_features=64, out_features=64, bias=True)
input = tensor([[[ 0.1404, -0.1627, -0.0633,  ..., -0.1266, -0.1406,  0.0955],
         [-0.0418,  0.1652,  0.1326,  ..., -0.1...58,  0.9727,  1.0664],
         [ 0.4693,  0.5153,  1.0113,  ..., -1.2772,  0.9666,  0.9658]]],
       device='cuda:0')

    def forward(self, input: Tensor) -> Tensor:
>       return F.linear(input, self.weight, self.bias)
E       RuntimeError: mat1 and mat2 shapes cannot be multiplied (162x100 and 64x64)
```

Any thoughts as to what's going on?

**Comment by patrickvonplaten** [comment]
> Hmm @patrickvonplaten I'm seeing a lot of failures in the fast pipeline tests using
> 
> ```python
> for component in pipe.components.values():
>     if hasattr(component, "set_default_attn_processor"):
>         component.set_default_attn_processor()
> ```
> 
> ```
> === 76 failed, 1314 passed, 741 skipped, 1469 warnings in 1015.78s (0:16:55) ===
> ```
> 
> They all seem to be coming from `lora.py`. Here's an example traceback
> 
> ```
> tests/pipelines/unclip/test_unclip_image_variation.py:480:
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/pipelines/test_pipelines_common.py:314: in test_save_load_local
>     output = pipe(**inputs)[0]
> /opt/venv/lib/python3.8/site-packages/torch/utils/_contextlib.py:115: in decorate_context
>     return func(*args, **kwargs)
> src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:314: in __call__
>     noise_pred = self.decoder(
> /opt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py:1501: in _call_impl
>     return forward_call(*args, **kwargs)
> src/diffusers/models/unet_2d_condition.py:930: in forward
>     sample, res_samples = downsample_block(
> /opt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py:1501: in _call_impl
>     return forward_call(*args, **kwargs)
> src/diffusers/models/unet_2d_blocks.py:1725: in forward
>     hidden_states = attn(
> /opt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py:1501: in _call_impl
>     return forward_call(*args, **kwargs)
> src/diffusers/models/attention_processor.py:419: in forward
>     return self.processor(
> src/diffusers/models/attention_processor.py:580: in __call__
>     key = attn.to_k(encoder_hidden_states, lora_scale=scale)
> /opt/venv/lib/python3.8/site-packages/torch/nn/modules/module.py:1501: in _call_impl
>     return forward_call(*args, **kwargs)
> src/diffusers/models/lora.py:117: in forward
>     return super().forward(hidden_states)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = LoRACompatibleLinear(in_features=64, out_features=64, bias=True)
> input = tensor([[[ 0.1404, -0.1627, -0.0633,  ..., -0.1266, -0.1406,  0.0955],
>          [-0.0418,  0.1652,  0.1326,  ..., -0.1...58,  0.9727,  1.0664],
>          [ 0.4693,  0.5153,  1.0113,  ..., -1.2772,  0.9666,  0.9658]]],
>        device='cuda:0')
> 
>     def forward(self, input: Tensor) -> Tensor:
> >       return F.linear(input, self.weight, self.bias)
> E       RuntimeError: mat1 and mat2 shapes cannot be multiplied (162x100 and 64x64)
> ```
> 
> Any thoughts as to what's going on?

Thanks for the ping, will give it a look today!

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion_img2img.py [hunk 8]
```diff
@@ -426,10 +426,8 @@ def run_safety_checker(self, image, device, dtype):
 
     def decode_latents(self, latents):
         warnings.warn(
-            (
-                "The decode_latents method is deprecated and will be removed in a future version. Please"
-                " use VaeImageProcessor instead"
-            ),
+            "The decode_latents method is deprecated and will be removed in a future version. Please"
+            " use VaeImageProcessor instead",
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion_img2img.py:430]:
```suggestion
            (
                "The decode_latents method is deprecated and will be removed in a future version. Please"
                " use VaeImageProcessor instead"
            ),
```

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py [hunk 9]
```diff
@@ -425,10 +425,8 @@ def run_safety_checker(self, image, device, dtype):
 
     def decode_latents(self, latents):
         warnings.warn(
-            (
-                "The decode_latents method is deprecated and will be removed in a future version. Please"
-                " use VaeImageProcessor instead"
-            ),
+            "The decode_latents method is deprecated and will be removed in a future version. Please"
+            " use VaeImageProcessor instead",
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:429]:
```suggestion
            (
                "The decode_latents method is deprecated and will be removed in a future version. Please"
                " use VaeImageProcessor instead"
            ),
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_img2img.py [hunk 10]
```diff
@@ -35,10 +35,8 @@
 # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion_img2img.preprocess with 8->64
 def preprocess(image):
     warnings.warn(
-        (
-            "The preprocess method is deprecated and will be removed in a future version. Please"
-            " use VaeImageProcessor.preprocess instead"
-        ),
+        "The preprocess method is deprecated and will be removed in a future version. Please"
+        " use VaeImageProcessor.preprocess instead",
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_img2img.py:39]:
```suggestion
        (
            "The preprocess method is deprecated and will be removed in a future version. Please"
            " use VaeImageProcessor.preprocess instead"
        ),
```

**Comment by patrickvonplaten** [comment]
Hey @DN6, 

I fixed the long list of failing tests. The remaining failing tests seem to be more related to precision

**Comment by DN6** [comment]
@patrickvonplaten  Fixed the failing tests. Merging this.

**Comment by patrickvonplaten** [comment]
Cool!

**Pull request merged**